In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_data_df = pd.read_csv(
    Path("Resources/lending_data.csv")
)

# Review the DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
x = lending_data_df.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
x.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [6]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(x, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [7]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [8]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)
print(len(testing_predictions))

19384


### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate a confusion matrix for the model
logistic_matrix = confusion_matrix(y_test, testing_predictions)
print(logistic_matrix)

[[18673    86]
 [   32   593]]


In [10]:
# Print the classification report for the model
logistic_classreport = classification_report(y_test, testing_predictions)
print(logistic_classreport)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.95      0.91       625

    accuracy                           0.99     19384
   macro avg       0.94      0.97      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Evaluation of Loan Prediction Model Performance

This analysis assesses a logistic regression model designed to predict loan status as either healthy (0) or high-risk (1). By analyzing key performance metrics, we can understand how well the model distinguishes between healthy and high-risk loans, assisting financial institutions in making informed, data-driven lending decisions.

Model Performance Summary

Accuracy Score: 0.99 (99%)
The model accurately classifies loan status in nearly all cases.

Precision:

Healthy Loans (0): 1.00 (100%)
Predictions for healthy loans are perfectly accurate, with a precision score of 100%.
High-Risk Loans (1): 0.84 (84%)
For high-risk loans, the model is 84% precise, meaning that when it flags a loan as high-risk, it is correct 84% of the time.
Recall:

Healthy Loans (0): 0.99 (99%)
The model successfully identifies 99% of actual healthy loans.
High-Risk Loans (1): 0.94 (94%)
It captures 94% of true high-risk loans, indicating strong sensitivity to risk detection.
Observations and Recommendations

The model is highly effective for predicting healthy loans, achieving 100% precision and 99% recall. With 18,765 healthy loans out of a total of 19,384, it reliably handles the majority class.
For high-risk loans, while performance remains strong (84% precision, 94% recall), it has a slightly lower precision. This means around 16% of loans flagged as high-risk may actually be healthy.
Given the imbalanced dataset, the model's ability to maintain high accuracy across both classes is commendable. However, to further refine loan assessments, consider:

Using the model as an initial screening tool while conducting a secondary review for high-risk loans flagged by the model.
Retraining the model regularly with updated data to adapt to changing loan characteristics and maintain accuracy over time.
In conclusion, this model provides reliable predictions, especially for healthy loans, and is a strong candidate for primary use in loan risk assessment.

---